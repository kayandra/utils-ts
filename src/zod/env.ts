import { z } from 'zod';

export type EnvOptionsType = {
  isServer?: boolean;
  clientPrefix?: string;
  skipValidation?: boolean;
};

export function createValidatedEnv<
  P extends Record<string, unknown>,
  R extends z.ZodRawShape,
>(payload: P, rawShape: R, options: EnvOptionsType = {}) {
  const validator = z.object(rawShape);
  const {
    skipValidation,
    clientPrefix = 'NEXT_PUBLIC_',
    isServer = typeof window === 'undefined',
  } = options;
  const validated = stripServerVars(
    skipValidation
      ? (payload as unknown as z.infer<typeof validator>)
      : validatePayload(payload, validator),
    clientPrefix,
    isServer,
  );

  return new Proxy(validated, {
    get(target, prop) {
      if (typeof prop !== 'string') return undefined;
      if (!isServer && !prop.startsWith(clientPrefix)) {
        throw new Error(
          `❌ Attempted to access server-side environment variable '${prop}' on the client`,
        );
      }

      return target[prop];
    },
  });
}

function validatePayload<
  P extends Record<string, unknown>,
  V extends z.ZodTypeAny,
>(payload: P, validator: V): z.infer<V> {
  const parsed = validator.safeParse(payload);
  if (parsed.success === false) {
    console.error(
      '❌ Invalid environment variables:',
      parsed.error.flatten().fieldErrors,
    );
    throw new Error('Invalid environment variables');
  }
  return parsed.data;
}

function stripServerVars<P extends Record<string, unknown>>(
  payload: P,
  prefix: string,
  isServer: boolean,
) {
  if (isServer) return payload;

  for (const [key] of Object.entries(payload)) {
    if (!key.startsWith(prefix)) {
      delete payload[key];
    }
  }

  return payload;
}
