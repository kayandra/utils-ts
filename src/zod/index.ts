export * from './env'
export * from './zod'
