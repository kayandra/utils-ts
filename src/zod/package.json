{
  "name": "@kydr/zod",
  "type": "module",
  "version": "1.0.0",
  "main": "index.ts",
  "types": "index.ts",
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "camelcase-keys": "^8.0.2",
    "type-fest": "^3.9.0"
  },
  "devDependencies": {
    "zod": "^3.21.4"
  },
  "peerDependencies": {
    "zod": "^3.21.4"
  }
}
