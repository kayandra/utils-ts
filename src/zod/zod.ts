import camelCaseKeys from 'camelcase-keys'
import { CamelCasedPropertiesDeep } from 'type-fest'
import * as z from 'zod'

export const propertiesToCamelCase = <T extends z.ZodTypeAny>(
  zod: T,
): z.ZodEffects<z.infer<T>, CamelCasedPropertiesDeep<T['_output']>> =>
  zod.transform<CamelCasedPropertiesDeep<T>>((val) => camelCaseKeys(val))
