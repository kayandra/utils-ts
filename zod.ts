import * as z from "zod";
import camelCaseKeys from "camelcase-keys";
import { CamelCasedPropertiesDeep } from "type-fest";

const zodToCamelCase = <T extends z.ZodTypeAny>(
  zod: T
): z.ZodEffects<z.infer<T>, CamelCasedPropertiesDeep<T["_output"]>> =>
  zod.transform((val) => camelCaseKeys(val) as CamelCasedPropertiesDeep<T>);
