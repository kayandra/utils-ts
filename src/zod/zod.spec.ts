import { describe, expect, it } from 'vitest';
import { z } from 'zod';

import { propertiesToCamelCase } from '.';

describe('Zod', () => {
  it('converts object keys to camelCase', () => {
    const schema = z.object({
      'kebab-case': z.string().default('kebabCase'),
      'SCREAMING-KEBAB-CASE': z.string().default('screamingKebabCase'),
      snake_case: z.string().default('snakeCase'),
      camel_Snake_Case: z.string().default('camelSnakeCase'),
      Pascal_Snake_Case: z.string().default('pascalSnakeCase'),
      SCREAMING_SNAKE_CASE: z.string().default('screamingSnakeCase'),
      PascalCase: z.string().default('pascalCase'),
    });

    const parsed = propertiesToCamelCase(schema).parse({});
    expect(parsed).toMatchSnapshot();
  });
});
