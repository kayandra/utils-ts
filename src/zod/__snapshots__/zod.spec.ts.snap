// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Zod > converts object keys to camelCase 1`] = `
{
  "camelSnakeCase": "camelSnakeCase",
  "kebabCase": "kebabCase",
  "pascalCase": "pascalCase",
  "pascalSnakeCase": "pascalSnakeCase",
  "screamingKebabCase": "screamingKebabCase",
  "screamingSnakeCase": "screamingSnakeCase",
  "snakeCase": "snakeCase",
}
`;
